src/audit/bypassable_contains_conditions.rs

```diff
@@ -0,0 +1,159 @@
+                } else {
+                    None
+                }
+            });
+
+        conditions
+            .flat_map(|(expr, loc)| {
+                Self::insecure_contains(expr).into_iter().map(move |(confidence, context)| {
+                    Self::finding()
+                        .severity(Severity::High)
+                        .confidence(confidence)
+                        .add_location(
+                            loc.with_keys(&["if".into()])
+                                .primary()
+                                .annotated(format!("contains(..) condition can be bypassed if attacker can control '{context}'")),
+                        )
+                        .build(job.parent())
+                })
+            })
+            .collect()
+    }
+}
+
+impl BypassableContainsConditions {
+    fn walk_tree_for_insecure_contains<'a>(
+        expr: &'a Expr,
+    ) -> Box<dyn Iterator<Item = (&'a str, &'a Context<'a>)> + 'a> {
+        match expr {
+            Expr::Call {
+                func: "contains",
+                args: exprs,
+            } => match exprs.as_slice() {
+                [Expr::String(s), Expr::Context(c)] => Box::new(std::iter::once((s.as_str(), c))),
+                args => Box::new(args.iter().flat_map(Self::walk_tree_for_insecure_contains)),
+            },
+            Expr::Call {
+                func: _,
+                args: exprs,
+            }
+            | Expr::Context(Context {
+                raw: _,
+                components: exprs,
+            }) => Box::new(exprs.iter().flat_map(Self::walk_tree_for_insecure_contains)),
+            Expr::Index(expr) => Self::walk_tree_for_insecure_contains(expr),
+            Expr::BinOp { lhs, rhs, .. } => {
+                let bc_lhs = Self::walk_tree_for_insecure_contains(lhs);
+                let bc_rhs = Self::walk_tree_for_insecure_contains(rhs);
+
+                Box::new(bc_lhs.chain(bc_rhs))
+            }
+            Expr::UnOp { expr, .. } => Self::walk_tree_for_insecure_contains(expr),
+            _ => Box::new(std::iter::empty()),
+        }
+    }
+
+    fn insecure_contains(expr: &str) -> Vec<(Confidence, String)> {
+        let bare = match ExplicitExpr::from_curly(expr) {
+            Some(raw_expr) => raw_expr.as_bare().to_string(),
+            None => expr.to_string(),
+        };
+
+        Expr::parse(&bare)
+            .inspect_err(|_err| tracing::warn!("couldn't parse expression: {expr}"))
+            .iter()
+            .flat_map(|expression| Self::walk_tree_for_insecure_contains(expression))
+            .map(|(_s, ctx)| {
+                let confidence = if USER_CONTROLLABLE_CONTEXTS.iter().any(|item| {
+                    let context = &ctx.as_str();
+                    item == context || (item.ends_with(".") && context.starts_with(item))
+                }) {
+                    Confidence::High
+                } else {
+                    Confidence::Medium
+                };
```

Comment:
Per above: I think this can always be `high` confidence, since it's a true finding either way -- we should instead vary the severity, since that's what user-controllable contexts signal 🙂